pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+from gcloud.aio.auth import BUILD_GCLOUD_REST
+
+if BUILD_GCLOUD_REST:
+    pass
+else:
+    import asyncio
+    import logging
+    import time
+    from typing import Awaitable
+    from typing import Callable
+    from typing import List
+    from typing import Optional
+    from typing import Tuple
+    from typing import TYPE_CHECKING
+
+    import aiohttp
+    from gcloud.aio.pubsub.subscriber_client import SubscriberClient
+    from gcloud.aio.pubsub.subscriber_message import SubscriberMessage
+    from gcloud.aio.pubsub.metrics_agent import MetricsAgent
+
+    log = logging.getLogger(__name__)
+
+    if TYPE_CHECKING:
+        MessageQueue = asyncio.Queue[Tuple[SubscriberMessage,  # pylint: disable=unsubscriptable-object
+                                           float]]
+    else:
+        MessageQueue = asyncio.Queue
+    ApplicationHandler = Callable[[SubscriberMessage], Awaitable[None]]
+
+    class AckDeadlineCache:
+        def __init__(self, subscriber_client: SubscriberClient,
+                     subscription: str, cache_timout: int):
+            self.subscriber_client = subscriber_client
+            self.subscription = subscription
+            self.cache_timeout = cache_timout
+            self.ack_deadline: float = float('inf')
+            self.last_refresh: float = 0
+
+        async def get(self) -> float:
+            if self.ack_deadline == float('inf') or self.cache_outdated():
+                await self.refresh()
+            return self.ack_deadline
+
+        async def refresh(self) -> None:
+            try:
+                sub = await self.subscriber_client.get_subscription(
+                    self.subscription)
+                self.ack_deadline = float(sub['ackDeadlineSeconds'])
+            except Exception as e:
+                log.warning(
+                    'Failed to refresh ackDeadlineSeconds value', exc_info=e)
+            self.last_refresh = time.perf_counter()
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
```

Comment:
@TheKevJames @allandialpad we were discussing if this is necessary
For: 
* will work regardless of subscription configuration management approach
Against:
* complexity

Reply:
I don't think anything has changed in my mind since last I commented; I continue to be inclined to mark this as unnecessary complexity; ideally we should be referring to a `gcloud.aio.pubsub.Subscription()['ackDeadLineSubscription']` object here and only listening for changes in cases where users do `gcloud.aio.pubsub.Subscription().patch(ackDeadlineSubscription=new_value)`.

Reply:
I'd vote for leaving the infrastructure here but just throwing a warning/error. Its possible that there is very common  -yet unused by us- usecase where apps dynamically modify their ack deadline

Reply:
@TheKevJames I still completely miss your point on patch. Do you mean `SubscriberClient` by `gcloud.aio.pubsub.Subscription()`? Could you throw some high level pseudo code here? If I drop refresh mechanism I need to replace it with something to get rid of possible bugs caused by desync and I don't see what less complex it could be.

Reply:
I was thinking something like the following, which is user-driven via gcloud-aio in app rather than requiring polling from the external service:
```python
class SubscriberClient:
    # SNIP
    def get_subscription(...) -> Subscription:
        # SNIP
        return Subscriber.from_repr(result)

class Subscriber:
    def from_repr(cls, data: Dict[str, Any]) -> 'Subscriber':
        # SNIP
        return Subscriber(..., ack_deadline_seconds=['ackDeadlineSeconds']

    # https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/patch
    def patch(self, ..., ack_deadline_seconds: Optional[int] = None) -> None:
        # SNIP
        if ack_deadline_seconds is not None:
            payload_body['ackDeadlineSeconds'] = ack_deadline_seconds
            self.ack_deadline_seconds = ack_deadline_seconds

async def consumer(subscription: Subscription, ...):
    while True:
        ack_deadline = subscription.ack_deadline_seconds
        # SNIP
```

The benefit there being that if users want to have their app code call `Subscription().patch(ack_deadline_seconds=42)`, it propagates to the running instance of the subscriber; but we don't need to worry about the extra logic of trying to keep our system in-sync with out-of-band changes.

Reply:
That's more complexity.
* we now store a local representation of a subscription
* we introduce a wrapper over `SubscriberClient`'s method => we now have 2 ways to patch subscription
* `Subscriber` will need to be updated on any schema change from google
* we require users to treat this representation as a the only agent of change for things to work properly
* unless there is only one instance running we now have N copies of the state, each of which is read-only-at-startup. The only way to keep them in sync is to restart everything on any write.

With `ackDeadlineCache` things will just work regardless of pretty much anything

I also want to point out that we're not polling _some_ external service. We're polling the one and the only external entity the subscriber is specifically designed to interact with.

Reply:
> unless there is only one instance running we now have N copies of the state, each of which is read-only-at-startup. The only way to keep them in sync is to restart everything on any write.

That's a good point; I didn't consider the multiple apps running simultaneously issue... though my base assumption was that folks would be scaling the same code horizontally in cases like that, so it would not be relevant. Bad assumption on my part!

> That's more complexity.

Yes and no -- it's more complexity in some ways, but I think it's more _useful_ complexity; eg. the complexity comes from defining the `Subscriber.patch()` API operation, which is probably something we'll want to do down the road _anyway_ as we find usecases for more of the Pubsub APIs. This solves the general class of "modify any subscription configs" rather than just the "respond to changes in ack deadline while subscribing case". It's always nice to try and avoid handling special cases specially when you can reframe the problem as it being the standard case.

That all said... fair enough, good points, I'm alright keeping this as-is.

Reply:
Haha so in the end we'll have both complexities, classical software engineers :D